### PR TITLE
[GPU] Refactored network/program/topology classes

### PR DESCRIPTION
**Pull request by vladimir-paramuzov**
### Details:
 - Removed pimpl idiom usage for graph classes
 - Build options were moved  to separate header
 - Added debug option for sources dump. Moved reading of DumpGraphs option to make it work in unit tests.

### CODE REVIEW

File: inference-engine/thirdparty/clDNN/src/network.cpp [hunk 1]
```diff
@@ -470,25 +389,36 @@ void network_impl::allocate_primitives() {
...
+
+    auto output_ids = get_output_ids();
+    std::map<primitive_id, network_output> result;
+    for (auto& id : output_ids) {
+        result.emplace(id, get_output(id));
+    }
+    return result;
+}
+
+void network::execute_impl(const std::vector<event::ptr>& events) {
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/src/network.cpp:421]: Wouldn't `execute_network` be a more appropriate name for the function?

File: inference-engine/thirdparty/clDNN/src/network.cpp [hunk 2]
```diff
@@ -470,25 +389,36 @@ void network_impl::allocate_primitives() {
...
     OV_ITT_SCOPED_TASK(itt::domains::CLDNN, "NetworkImpl::Execute");
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/src/network.cpp:422]: I think it makes sense to modify ITT marks according to relevant functions names

**Review by sshlyapn — APPROVED** [review]
In general looks good to me

File: inference-engine/thirdparty/clDNN/src/network.cpp (hunk 1 above)
Comment (vladimir-paramuzov) [reply inference-engine/thirdparty/clDNN/src/network.cpp:421]: I think `network::execute_impl` is better than `network::execute_network`

File: inference-engine/thirdparty/clDNN/src/network.cpp (hunk 2 above)
Comment (vladimir-paramuzov) [reply inference-engine/thirdparty/clDNN/src/network.cpp:422]: Agree, but it's not in a scope of current PR. Let's do it separately

File: inference-engine/thirdparty/clDNN/api/cldnn/graph/program.hpp [hunk 3]
```diff
@@ -5,501 +5,324 @@
...
-    std::vector<std::shared_ptr<const build_option>> _options;
-    void set_option(void) {}
-
-    void add_or_replace_option(std::shared_ptr<const build_option> opt) {
-        for (auto& p : _options) {
-            if (p->get_type() == opt->get_type()) {
-                p = opt;
-                return;
-            }
+        // int32_t get_processing_number(const_iterator iter) const { return 1+(int32_t)std::distance(begin(), iter); }
```
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply inference-engine/thirdparty/clDNN/api/cldnn/graph/program.hpp]: done

**Pull request merged**